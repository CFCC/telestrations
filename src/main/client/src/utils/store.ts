import { createGamekitSlice, sendEvent } from "@piticent123/gamekit-client";
import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import { GameState, TelestrationsGame, TelestrationsPlayer } from "./types";
import _ from "lodash";
import { Settings } from "@piticent123/gamekit-client/lib/types";

const { reducer: gamekitReducer, actions: gamekitActions } =
  createGamekitSlice<Settings, TelestrationsPlayer, TelestrationsGame>();

interface AppState {
  gameState: GameState;
  activeContentId: string;
}
interface State {
  app: AppState;
  gamekit: ReturnType<typeof gamekitReducer>;
}
interface ThunkApi {
  state: State;
}

// This has to be at the top level to be throttled correctly
const _sendWritePageEvent = _.throttle((gameCode: string, content: string) => {
  sendEvent({
    route: `/app/games/${gameCode}/page/write`,
    data: content,
  });
}, 500);
export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    _sendWritePageEvent(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({
      route: `/app/games/${currentGame.code}/page/submit`,
      data: content,
    });
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (__, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({ route: `/games/${currentGame.code}/start` });
  }
);

export const {
  createGame,
  joinGame,
  rejoinGame,
  connectToServer,
  saveSettings,
  becomeAdmin,
} = gamekitActions;

const { actions, reducer: appReducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    activeContentId: "",
  } as AppState,
  reducers: {
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(gamekitActions.rejoinGame.fulfilled, (state) => {
        state.gameState = GameState.IN_GAME;
      })
      .addCase(gamekitActions.connectToServer.fulfilled, (state) => {
        state.gameState = GameState.GAME_CODE;
      })
      .addCase(gamekitActions.joinGame.fulfilled, (state) => {
        state.gameState = GameState.WAITING_TO_START;
      })
      .addCase(submitGuess.fulfilled, (state) => {
        state.gameState = GameState.WAITING_FOR_CONTENT;
      });
  },
});

export const {
  viewNotepadHistory,
  viewPlayerHistory,
  setGameState
} = actions;

export const store = configureStore({
  reducer: { gamekit: gamekitReducer, app: appReducer },
});

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
