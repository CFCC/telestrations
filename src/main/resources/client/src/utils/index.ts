export function sleep(t: number): Promise<void> {
  return new Promise((res) => setTimeout(res, t));
}
