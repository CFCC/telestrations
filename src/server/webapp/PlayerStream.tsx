import React, {useContext} from "react";
import {CardContent, Typography} from "@material-ui/core";
import styled from "styled-components";
import {UUID} from "types/shared";
import {GameContext} from "server/webapp/Store";

interface PlayerStreamProps {
    playerId: UUID;
}

const Picture = styled.img`
    max-width: 100%;
    min-width: 100%;
    height: auto;
    box-shadow: inset 0 2px 4px 0 hsla(0, 0%, 0%, 0.2);
`;

export default function PlayerStream({playerId}: PlayerStreamProps) {
    const [{players, notepads}] = useContext(GameContext);
    const playerIndex = players.findIndex(p => p.id === playerId);
    const notepadIndex = notepads.findIndex(n => n.owner === players[playerIndex].ownerOfCurrentNotepad);

    const playerIndexInNotepad = players[playerIndex].notepadIndex;
    const playerDrawing = playerIndexInNotepad % 2 === 0;
    const playerWaiting = playerIndexInNotepad === -1;

    const prevContent = playerIndexInNotepad === 0 || playerWaiting
        ? "/question-marks.jpg"
        : notepads[notepadIndex].content[playerIndexInNotepad - 1];
    const content = playerWaiting
        ? "Waiting for next notepad..."
        : notepads[notepadIndex].content[playerIndexInNotepad];

    let picture;
    if (playerWaiting || playerIndexInNotepad === 0) picture = "/question-marks.jpg";
    else if (playerDrawing) picture = `http://localhost:${process.env.REACT_APP_SERVER_PORT}${prevContent}`;
    else picture = `http://localhost:${process.env.REACT_APP_SERVER_PORT}${content}`;

    return (<React.Fragment>
        <Picture
            src={picture}
            alt={playerDrawing || playerWaiting ? content : prevContent}
        />
        <CardContent>
            <Typography align="center">
                {playerDrawing || playerWaiting ? content : prevContent}
            </Typography>
        </CardContent>
    </React.Fragment>);
}
