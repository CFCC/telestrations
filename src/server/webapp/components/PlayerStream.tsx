import React, {useContext} from "react";
import {withStyles} from "@material-ui/core";
import {ClassProps, UUID} from "../../../types/shared";
import {GameContext} from "../Store";

interface PlayerStreamProps extends ClassProps {
    playerId: UUID;
}

export default withStyles({
    app: {},
})(function PlayerStream({classes, playerId}: PlayerStreamProps) {
    const [{players, notepads}] = useContext(GameContext);
    const playerIndex = players.findIndex(p => p.id === playerId);
    const notepadIndex = notepads.findIndex(n => n.owner === players[playerIndex].ownerOfCurrentNotepad);

    const prevContent = players[playerIndex].notepadIndex === 0
        ? "/question-marks.jpg"
        : notepads[notepadIndex].content[players[playerIndex].notepadIndex - 1];
    const content = notepads[notepadIndex].content[players[playerIndex].notepadIndex];

    return players[playerIndex].notepadIndex % 2 === 0 ? <div className={classes.app}>
        <img src={prevContent} alt={content} />
        <h5>{content}</h5>
    </div> : <div className={classes.app}>
        <img src={content} alt={prevContent} />
        <h5>{prevContent}</h5>
    </div>;
});
