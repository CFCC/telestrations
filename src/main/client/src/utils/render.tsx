import React from "react";
import { render } from "@testing-library/react";
import { StylesProvider, ThemeProvider } from "@material-ui/core/styles";
import { Provider } from "react-redux";

import { GlobalStyles, theme } from "./theme";
import { store } from "./store";

interface WrapperProps {
  children: React.ReactNode;
}

const wrapper = ({ children }: WrapperProps) => (
  <React.Fragment>
    <StylesProvider injectFirst={true}>
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Provider store={store}>{children}</Provider>
      </ThemeProvider>
    </StylesProvider>
  </React.Fragment>
);
const customRender = (ui: React.ReactElement) =>
  render(ui, { wrapper } as object);

export { customRender as render };
