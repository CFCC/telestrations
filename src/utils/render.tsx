import React from "react";
import {render} from "@testing-library/react";
import {StylesProvider, ThemeProvider} from "@material-ui/core/styles";
import {Provider} from "react-redux";

import {GlobalStyles, theme} from "./theme";
import {store} from "../store/client";

interface WrapperProps {
    children: React.ReactNode;
}

const wrapper = (store: any) => ({children}: WrapperProps) => (
    <React.Fragment>
        <StylesProvider injectFirst={true}>
            <ThemeProvider theme={theme}>
                <GlobalStyles />
                <Provider store={store}>
                    {children}
                </Provider>
            </ThemeProvider>
        </StylesProvider>
    </React.Fragment>
);
const customRender = (ui: React.ReactElement, options: any) => render(ui, {wrapper, ...options});

export {customRender as render};
