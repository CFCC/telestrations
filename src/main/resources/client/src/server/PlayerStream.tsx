import React, {useEffect, useState} from "react";
import {CardContent, Typography} from "@material-ui/core";
import styled from "styled-components";

import {useSelector} from "../utils/store";
import {getImageURL} from "../utils/firebase";

interface PlayerStreamProps {
    playerId: string;
}

interface PictureContainerProps {
    picture: string;
}

const PictureContainer = styled.div`
    background-image: ${(props: PictureContainerProps) => props.picture === "/question-marks.jpg" ? "url(/question-marks.jpg)" : "none"};
    background-repeat: repeat;
    width: 100%;
    min-height: ${(props: PictureContainerProps) => props.picture === "/question-marks.jpg" ? "300px" : "0"};
`;

const Picture = styled.img`
    max-width: 100%;
    min-width: 100%;
    height: auto;
`;

const Content = styled(CardContent)`
    padding-bottom: 0;
`;

export default function PlayerStream({playerId}: PlayerStreamProps) {
    const {players, notepads} = useSelector(state => state.firebase);
    const [pictureUrl, setPictureUrl] = useState("");

    const notepad = notepads[players[playerId]?.currentNotepad];
    const playerIndexInNotepad = notepad?.pages?.length;

    let picture: string, t, text;
    if (playerIndexInNotepad == null) {
        picture = "/question-marks.jpg";
        text = "Waiting for next notepad...";
    } else if (playerIndexInNotepad <= 1) {
        picture = "/question-marks.jpg";
        text = notepad.pages[playerIndexInNotepad - 1]?.content;
    } else if (playerIndexInNotepad % 2 === 0) {
        picture = notepad.pages[playerIndexInNotepad - 1]?.content;
        t = notepad.pages[playerIndexInNotepad - 1]?.lastUpdated;
        text = notepad.pages[playerIndexInNotepad - 2]?.content;
    } else {
        picture = notepad.pages[playerIndexInNotepad - 2]?.content;
        t = notepad.pages[playerIndexInNotepad - 2]?.lastUpdated;
        text = notepad.pages[playerIndexInNotepad - 1]?.content;
    }

    useEffect(() => {
        (async function() {
            const url = await getImageURL(picture);
            setPictureUrl(url);
        })();
    }, [picture, t]);

    return (
        <React.Fragment>
            <PictureContainer picture={picture}>
                {picture !== "/question-marks.jpg" && (
                    <Picture
                        src={pictureUrl}
                        alt={text}
                    />
                )}
            </PictureContainer>
            <Content>
                <Typography align="center">{text}</Typography>
            </Content>
        </React.Fragment>
    );
}
