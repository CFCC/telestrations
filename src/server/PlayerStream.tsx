import React, {useContext} from "react";
import {CardContent, Typography} from "@material-ui/core";
import styled from "styled-components";
import {GameContext} from "../store/server";

interface PlayerStreamProps {
    playerId: string;
}

interface PictureContainerProps {
    picture: string;
}

const PictureContainer = styled.div`
    background-image: ${(props: PictureContainerProps) => props.picture === "/question-marks.jpg" ? "url(/question-marks.jpg)" : "none"};
    background-repeat: repeat;
    width: 100%;
    min-height: ${(props: PictureContainerProps) => props.picture === "/question-marks.jpg" ? "300px" : "0"};
`;

const Picture = styled.img`
    max-width: 100%;
    min-width: 100%;
    height: auto;
`;

const Content = styled(CardContent)`
    padding-bottom: 0;
`;

export default function PlayerStream({playerId}: PlayerStreamProps) {
    const [{game: {players, notepads}}] = useContext(GameContext);

    const notepad = notepads[players[playerId]?.currentNotepad];
    const playerIndexInNotepad = notepad?.pages?.length;

    let picture, text;
    if (playerIndexInNotepad == null) {
        picture = "/question-marks.jpg";
        text = "Waiting for next notepad...";
    } else if (playerIndexInNotepad <= 1) {
        picture = "/question-marks.jpg";
        text = notepad.pages[playerIndexInNotepad - 1]?.content;
    } else if (playerIndexInNotepad % 2 === 0) {
        picture = notepad.pages[playerIndexInNotepad - 1]?.content;
        text = notepad.pages[playerIndexInNotepad - 2]?.content;
    } else {
        picture = notepad.pages[playerIndexInNotepad - 2]?.content;
        text = notepad.pages[playerIndexInNotepad - 1]?.content;
    }

    return (
        <React.Fragment>
            <PictureContainer picture={picture}>
                {picture !== "/question-marks.jpg" && (
                    <Picture
                        src={picture}
                        alt={text}
                    />
                )}
            </PictureContainer>
            <Content>
                <Typography align="center">{text}</Typography>
            </Content>
        </React.Fragment>
    );
}
