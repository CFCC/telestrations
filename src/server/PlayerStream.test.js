import {render} from "../utils/render";
import {WhileDrawing, WhileTyping} from "./PlayerStream.stories";

it('While Drawing Renders', () => {
    const {asFragment} = render(WhileDrawing());
    expect(asFragment()).toMatchSnapshot();
});

it('While Typing Renders', () => {
    const {asFragment} = render(WhileTyping());
    expect(asFragment()).toMatchSnapshot();
});

