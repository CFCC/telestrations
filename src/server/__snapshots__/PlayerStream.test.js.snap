// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`While Drawing Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm ehZnxG"
  />
  <div
    class="MuiCardContent-root sc-AxiKw fiNxug"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
    >
      Drawing for next notepad...
    </p>
  </div>
</DocumentFragment>
`;

exports[`While Typing Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm ehZnxG"
  />
  <div
    class="MuiCardContent-root sc-AxiKw fiNxug"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
    >
      Drawing for next notepad...
    </p>
  </div>
</DocumentFragment>
`;
